src/debug_sdfs_ed64.c

```diff
@@ -0,0 +1,590 @@
+}
+
+void everdrive_sd_busy() {
+	while ((io_read(ED64_BASE_ADDRESS + ED64_REGISTER_SD_STATUS)
+		& ED64_SD_STATUS_BUSY) != 0);
+}
+
+void everdrive_sd_write_command(uint8_t val) {
+	everdrive_sd_set_mode(ED64_SD_MODE_CMD_WRITE);
+	io_write(ED64_BASE_ADDRESS + ED64_REGISTER_SD_BASE + ED64_SD_MODE_CMD_WRITE, val);
+	everdrive_sd_busy();
+}
+
+uint8_t everdrive_sd_read_command() {
+	everdrive_sd_set_mode(ED64_SD_MODE_CMD_READ);
+	io_write(ED64_BASE_ADDRESS + ED64_REGISTER_SD_BASE + ED64_SD_MODE_CMD_READ, 0xffff);
+	everdrive_sd_busy();
+	return io_read(ED64_BASE_ADDRESS + ED64_REGISTER_SD_BASE + ED64_SD_MODE_CMD_READ);
+}
+
+void everdrive_sd_write_data(uint8_t val) {
+	everdrive_sd_set_mode(ED64_SD_MODE_DATA_WRITE);
+	io_write(ED64_BASE_ADDRESS + ED64_REGISTER_SD_BASE + ED64_SD_MODE_DATA_WRITE,  0x00ff | (val << 8));
+}
+
+uint8_t everdrive_sd_read_data() {
+	everdrive_sd_set_mode(ED64_SD_MODE_DATA_READ);
+	io_write(ED64_BASE_ADDRESS + ED64_REGISTER_SD_BASE + ED64_SD_MODE_DATA_READ,  0xffff);
+	return io_read(ED64_BASE_ADDRESS + ED64_REGISTER_SD_BASE + ED64_SD_MODE_DATA_READ);
+}
+
+// Wait for and read the first byte
+uint8_t everdrive_sd_read_first() {
+	uint8_t res;
+	uint32_t timeout = 64; // 8 bytes * 8 bits
+
+	set_everdrive_sd_bitlen(8);
+	res = everdrive_sd_read_command();
+
+	// Effectively we are bitshifting the command buffer until we find a zero
+	// (start bit), followed by another zero (transmission bit)
+	// We should be able to find it in 8 bytes
+	set_everdrive_sd_bitlen(1);
+	while ((res & 0x80) != 0) {
+		if (!timeout--) return 0xFF;
+		res = everdrive_sd_read_command();
+	}
+	return res;
+}
+
+static uint8_t everdrive_sd_resp_buff[5] = { 0 };
+// Wait for and read an RX like response
+static uint8_t everdrive_sd_read_RX() {
+	uint8_t timeout = 16;
+	everdrive_sd_resp_buff[0] = everdrive_sd_read_first();
+
+	// read remaining bytes
+	set_everdrive_sd_bitlen(8);
+	everdrive_sd_resp_buff[1] = everdrive_sd_read_command();
+	everdrive_sd_resp_buff[2] = everdrive_sd_read_command();
+	everdrive_sd_resp_buff[3] = everdrive_sd_read_command();
+	everdrive_sd_resp_buff[4] = everdrive_sd_read_command();
+
+	// Make sure everything is consumed, we just don't use them
+	while(everdrive_sd_read_command() != 0xFF) {
+		if (!timeout--) return 0xFF;
+	};
+	return everdrive_sd_resp_buff[0];
+}
+
+static uint8_t everdrive_sd_execute_command(uint8_t cmd, uint32_t arg) {
+	uint64_t crc7 = 0x00; // Most significant byte will be the result
+
+	crc7 = (uint64_t)cmd << 56;
+	crc7 |= (uint64_t)arg << 24;
+	for (int i = 0; i < 40; i++) {;
+		uint64_t hibit = crc7 >> 63;
+		crc7 <<= 1;
+		if (hibit) crc7 = (crc7 ^ ((uint64_t)0x12 << 56));
+	}
+
+	set_everdrive_sd_bitlen(8);
+
+	everdrive_sd_write_command(0xff);
+	everdrive_sd_write_command(cmd);
+	everdrive_sd_write_command(arg >> 24);
+	everdrive_sd_write_command(arg >> 16);
+	everdrive_sd_write_command(arg >> 8);
+	everdrive_sd_write_command(arg);
+
+	// LS bit must always be 1
```

Comment:
I'm not used to the notation `LS bit`. I read that like there is a bit called "LS" in the reference manual (line signal, or whatever). For least-significant bit, I usually see "LSB", or "bit 0".

Reply:
Done